ding/world_model/utils.py

```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
+        self.probs = torch.softmax(logits, -1)
+        self.buckets = torch.linspace(low, high, steps=255).to(device)
+        self.width = (self.buckets[-1] - self.buckets[0]) / 255
+
+    def mean(self):
+        print("mean called")
+        _mean = self.probs * self.buckets
+        return symexp(torch.sum(_mean, dim=-1, keepdim=True))
+
+    def mode(self):
+        _mode = self.probs * self.buckets
+        return symexp(torch.sum(_mode, dim=-1, keepdim=True))
+
+    # Inside OneHotCategorical, log_prob is calculated using only max element in targets
+    def log_prob(self, x):
+        x = symlog(x)
+        # x(time, batch, 1)
+        below = torch.sum((self.buckets <= x[..., None]).to(torch.int32), dim=-1) - 1
+        above = len(self.buckets) - torch.sum((self.buckets > x[..., None]).to(torch.int32), dim=-1)
+        below = torch.clip(below, 0, len(self.buckets) - 1)
+        above = torch.clip(above, 0, len(self.buckets) - 1)
+        equal = (below == above)
+
+        dist_to_below = torch.where(equal, 1, torch.abs(self.buckets[below] - x))
+        dist_to_above = torch.where(equal, 1, torch.abs(self.buckets[above] - x))
+        total = dist_to_below + dist_to_above
+        weight_below = dist_to_above / total
+        weight_above = dist_to_below / total
+        target = (
+            F.one_hot(below, num_classes=len(self.buckets)) * weight_below[..., None] +
+            F.one_hot(above, num_classes=len(self.buckets)) * weight_above[..., None]
+        )
+        log_pred = self.logits - torch.logsumexp(self.logits, -1, keepdim=True)
+        target = target.squeeze(-2)
+
+        return (target * log_pred).sum(-1)
+
+    def log_prob_target(self, target):
+        log_pred = super().logits - torch.logsumexp(super().logits, -1, keepdim=True)
+        return (target * log_pred).sum(-1)
+
+
+class SymlogDist():
+
+    def __init__(self, mode, dist='mse', agg='sum', tol=1e-8, dim_to_reduce=[-1, -2, -3]):
+        self._mode = mode
+        self._dist = dist
+        self._agg = agg
+        self._tol = tol
+        self._dim_to_reduce = dim_to_reduce
+
+    def mode(self):
+        return symexp(self._mode)
+
+    def mean(self):
+        return symexp(self._mode)
+
+    def log_prob(self, value):
+        assert self._mode.shape == value.shape
+        if self._dist == 'mse':
+            distance = (self._mode - symlog(value)) ** 2.0
+            distance = torch.where(distance < self._tol, 0, distance)
+        elif self._dist == 'abs':
+            distance = torch.abs(self._mode - symlog(value))
+            distance = torch.where(distance < self._tol, 0, distance)
+        else:
+            raise NotImplementedError(self._dist)
+        if self._agg == 'mean':
+            loss = distance.mean(self._dim_to_reduce)
+        elif self._agg == 'sum':
+            loss = distance.sum(self._dim_to_reduce)
+        else:
+            raise NotImplementedError(self._agg)
+        return -loss
+
+
+class ContDist:
```

Comment:
polish all the class name of distributions we implemented, unify their name format